Fold broadcasts in `GEMM([g,m,k], [broadcast(g), k,n])` on `m`

giuseros opened the pull request:
This is the pass I wrote to fold the broadcast on the batch to the `M` dimension. 

IR before:
```
A = <GxMxK>
B = <broadcast{1}(G)xKxN>
C = gemm(A, B)
```
IR after:
```
A = <G*MxK>
B = <KxN>
C = gemm(A, B)
```

I discussed with @manupak the possibility to compute blockSize and gridSize just before `gemmToGridwise` and created the ticket https://github.com/ROCm/rocMLIR-internal/issues/1319. For now I get away by re-running the `affixTuningParameters` pass once more after broadcasts dimensions are folded.

giuseros committed:
Fold broadcasts in `GEMM([g,m,k], [broadcast(g), k,n])` on `m`

manupak commented on mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:
```diff
@@ -0,0 +1,174 @@
+//===- Regularize.cpp - rewrites to allow Rock kernel fusion  ------===//
```
copy-pasta :)

manupak commented on mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:
```diff
@@ -0,0 +1,174 @@
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-canonicalize-gemm"
```
rock-fold-broadcast ?

manupak commented on mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:
```diff
@@ -0,0 +1,174 @@
+    // Look for a `Broadcast{1} at [0]` transformation
+    for (auto op : ops)
+      if (op.getType() == rock::TransformType::Broadcast &&
+          op.getUpperDims().back() == 0 && op.getParams().back() == 1)
```
Im a bit vary about this check... it is a bit fragile because unit dims / reshapes / transposes ? can easily break the ordering but it could be a still be a isBatchDimFoldable.

I think we need to backtrack a stride calculation on the batch dimension to see if its end up being zero on the actual memory layout.

giuseros replied:
Just to be sure, you are saying that a broadcast might also not be upfront, right? It might be a passthrough/slice that leads to a broadcast.

manupak replied:
yes -- any other transform to be precise.

giuseros replied:
I am not sure what you mean with "any other transform"? I was thinking to "rerouting" transforms(like passthrough), because I am not sure how can I deal with `broadcast` buried behind a Merge/unmerge/embed

manupak replied:
well the idea I had was to calculate stride of "g" dimension traversing up the trMaps.
Cant we handle the stride calculation in a merge/unmerge as the cumalative mulplication of faster changing dimensions in the merge set ? 

Embed is the tricky one but since we are only interested of figuring out stride is zero or not. We can possibly inherit the stride from bottom coords from the embed set

manupak replied:
I think it is quite plausible that we may encounter shape collapsing and expansion around the broadcast -- which usually result in merge/unmerge 

Embeds occur in conv to gemm I think. Thats a good point! we should test a convolution too.

krzysz00 replied:
I think the test is:

Follow the `G` dimension (this'll be an index, starts at 0) while looping through the transform stack.

PassThrough updates the index, as does Slice (given what condition we're looking for).

If, at some point, the transform that produced the current position of the G dimension is a Broadcast of a length-1 dimension, this is a foldable broadcast. Alternatively, if the G dimension is AddDim{}'d on, that's also a broadcast and so passes the check.

If the G dimension is created by Merge[*], Unmerge, or Embed, abort.

[*] There's a possible exception for Merge{1, 1, len(G), 1, 1}-type cases, but that's Effort and unlikely - unless we find it an example. Same goes for trivial Unmerge operations.

manupak replied:
That can also work in most cases until batches are collapsed/expanded after the broadcast -- which I agree is unlikely.
I was more speaking about strides thing was : dont we need that for problem key v2 ?. 
Therefore if we have that, it is just a matter of looking at the stride of "g".

giuseros replied:
Yep this was a plan I had, but I didn't do it because I didn't know if I was being too general. I think my confusion is about: 

> If the G dimension is created by Merge[*], Unmerge, or Embed, abort.

@manupak is that what you meant? Or you were suggesting to go through these transformations as well?

manupak replied:
So there is a reason we have the following code (it was done by Jungwook): 

https://github.com/ROCm/rocMLIR/blob/3202ef9b92a9278708b445a2611fff41750ffeea/mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp#L297-L321

to me it indicates that we will probably add a reshape to batch dimension and it would be after the broadcast because the above case handle where there are multiple dimensions (can just be unit dims) and fold them to be a single batch dimension.

manupak replied:
> @manupak is that what you meant? Or you were suggesting to go through these transformations as well?

I was suggesting to calculate/update a stride as we go through the stack and stop if and when it becomes zero (i.e. broadcast).

I ll point out that @krzysz00 's logic will work until we have a reshape that involves batch dimension after the broadcast. Now, thinking about it in the broadcast case, we might not commonly exercise that case that but I am a bit worried about unit dims.

Anyway, when we have problem key v2, we will have to calculate the strides. Therefore, whatever logic we are discussing could just go away and we can just take the decision looking at the stride ?

manupak commented on mlir/test/fusion/pr-e2e/mixr-gemm/mixr-gemm-broadcast.e2d.mlir:
```diff
@@ -0,0 +1,19 @@
+// RUN: rocmlir-driver -kernel-pipeline migraphx,highlevel %s | rocmlir-gen -ph -print-results -rand none - | rocmlir-driver -arch %arch -c  | mlir-cpu-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_float16_utils%shl ...
```
Here is an example I d like to see working : 

```mlir
func.func @mlir_transpose_slice_dot_add_add_tanh_add_sigmoid_sub_mul_mul_add_real(%arg0: !migraphx.shaped<1x64xf32, 64x1>, %arg1: !migraphx.shaped<32x64xf32, 64x1>, %arg2: !migraphx.shaped<32x64xf32, 64x1>, %arg3: !migraphx.shaped<32x64xf32, 64x1>, %arg4: !migraphx.shaped<32x64xf32, 64x1>, %arg5: !migraphx.shaped<32x64xf32, 64x1>, %arg6: !migraphx.shaped<32x64xf32, 64x1>, %arg7: !migraphx.shaped<128x64xf32, 64x1>) -> !migraphx.shaped<32x64xf32, 64x1> attributes {kernel = "mixr", num_cu = 48 : i64} {
    %0 = migraphx.multibroadcast %arg0 {out_dyn_dims = [], out_lens = [32, 64]} : <1x64xf32, 64x1> -> <32x64xf32, 0x1>
    %1 = migraphx.transpose %arg7 {permutation = [1, 0]} : <128x64xf32, 64x1> -> <64x128xf32, 128x1>
    %2 = migraphx.slice %1 {axes = [1], ends = [128], starts = [64]} : <64x128xf32, 128x1> -> <64x64xf32, 64x1>
    %3 = migraphx.dot %arg6, %2 : <32x64xf32, 64x1>, <64x64xf32, 64x1> -> <32x64xf32, 64x1>
    %4 = migraphx.add %3, %0 : <32x64xf32, 64x1>, <32x64xf32, 0x1> -> <32x64xf32, 64x1>
    %5 = migraphx.add %arg1, %4 : <32x64xf32, 64x1>, <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    %6 = migraphx.tanh %5 : <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    %7 = migraphx.add %arg2, %arg3 : <32x64xf32, 64x1>, <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    %8 = migraphx.sigmoid %7 : <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    %9 = migraphx.sub %arg4, %8 : <32x64xf32, 64x1>, <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    %10 = migraphx.mul %9, %6 : <32x64xf32, 64x1>, <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    %11 = migraphx.mul %8, %arg5 : <32x64xf32, 64x1>, <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    %12 = migraphx.add %10, %11 : <32x64xf32, 64x1>, <32x64xf32, 64x1> -> <32x64xf32, 64x1>
    return %12 : !migraphx.shaped<32x64xf32, 64x1>
  }
```

krzysz00 replied:
@manupak Can you clarify how the optimization we're discussing applies to that example? I don't see it here.

manupak replied:
Yea -- it was not related to the optimization but just wanted to highlight this is also a foldable broadcast.

an example could be -- also that would also work with your logic (the traversal you mentioned earlier)

128 x f32
 |
V
4 x 32 x f32
 | bcast
V
16 x 4 x 32 x f32

manupak replied:
ok Im just realizing there is no batch dimension at all. sigh. nvm

manupak replied:
anyways, my ask was to add something that does not have 'just'  a multibroadcast.

manupak replied:
After having a second thought, this is actually a counter-example where there is a broadcast on the first dimension where we should *NOT* fold.

manupak reviewed (commented):
I think we need to do a bit more than tracking down a broadcast.
I have given a *suggestion* but happy to hear alternatives, however, I d like us to nail down the issue.

manupak commented on mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:
```diff
@@ -0,0 +1,174 @@
+    if (views.empty())
+      return false;
+
+    auto trMap = views[0].cast<TransformMapAttr>();
```
It also can be on any views[N] not neccesarily on views[0]

krzysz00 commented on mlir/include/mlir/Dialect/Rock/IR/RockGemmWrapperInterface.td:
```diff
@@ -156,6 +156,18 @@ def RockGemmWrapperInterface : OpInterface<"RockGemmWrapperInterface"> {
           $_op->setAttr($_op.getParamsAttrName(), params);
         }]
       >,
+    InterfaceMethod<
```
Didn't we factor this out into another interface? Or is that something we just planned to do?

krzysz00 replied:
Looks like it was just planned

krzysz00 commented on mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp:
```diff
@@ -124,14 +124,18 @@ void rock::buildKernelPipeline(OpPassManager &pm,
                                                  options.tuningFallback}));
   funcPm.addPass(rock::createRockConvToGemmPass());
+  funcPm.addPass(rock::createRockFoldBroadcastPass());
+  funcPm.addPass(rock::createRockAffixTuningParametersPass(
```
Could you run this conditionally from within `createRockFoldBroadcastPass()`?

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:
```diff
@@ -0,0 +1,174 @@
+//===- Regularize.cpp - rewrites to allow Rock kernel fusion  ------===//
+//
+// Copyright 2022 Advanced Micro Devices.
```
Maybe also fix date?

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:
```diff
@@ -0,0 +1,174 @@
+    rock::TopDownTMBuilder unbroadcastBuilder(rw, {"d0", "d1"},
+                                              {shape[1], shape[2]}, loc);
+    unbroadcastBuilder.constDim({"g"}, 0, 0, shape[0]);
+    unbroadcastBuilder.passThrough({"d0"}, {1}, {"d0"});
```
Nit: you can combine the two `passThrough` calls

giuseros committed:
Move the pass into highlevel and generalize the broadcast check

giuseros commented:
Ok, so I updated the pass :
- It now runs in the `highlevel` pipeline
- It now follows the dimensions through their transformations (I added also more complicated tests)

giuseros commented:
@manupak @krzysz00  a gentle reminder :)

krzysz00 reviewed (approved):
This all looks good to me

manupak reviewed (approved):
Looks great now!
